Move to TS using Ackama Serverless Template as base

eoinkelly opened the pull request:
Closes #22
Closes #23
Closes #24

Changes

* Convert to the Ackama serverless template
* Convert JS to TS
* Fix bug where date conversions depended on system TZ
* Replace all use of `date-fns` with `luxon`
* Fix bug where we error sending to Slack because message has 50+ blocks. Now we send multiple messages of 50 blocks if required (fixes issue #24)

eoinkelly committed:
Recreate this app on the Ackama serverless template

eoinkelly committed:
Fix how sls invokes the handlers

eoinkelly committed:
Skip TS checks on all existing JS files to allow slower adoption

eoinkelly committed:
Prettify serverless.yml

eoinkelly committed:
Remove dependency on date-fns and convert some stuff to TS

eoinkelly committed:
Move files to TS

eoinkelly committed:
Fix some eslint errors

eoinkelly committed:
Fix last eslint warnings

eoinkelly committed:
Don't depend on system TZ for date math

eoinkelly committed:
Improve type narrowing in projectworks code

eoinkelly committed:
Docs clean-up

eoinkelly commented on src/publicHolidayReminder.ts:
```diff
@@ -41,33 +48,53 @@ const toDateStamp = date => {
+  return item.type === 'VEVENT';
+};
+
+interface MyEvent {
+  start: DateWithTimeZone;
+  description: string;
+}
```
I put most types in `types.ts` but for some types which are only used within a single module I have left them in the module. Not sure if this is the right call.

G-Rath replied:
imo the important things are to encapsulate the types based on their namespace/coupling, and that you don't create unneeded import/exports by pulling out every type even if it's only being used by one module.

For this project I think what you've got is pretty fine, since I don't think there's enough actual code to justify any extra namespaces but if there was a bit more to it I'd say you should maybe think about having `projectworks` and `notifier` namespaces in which case their types related types should live in there too.

eoinkelly replied:
Thanks. Good to know where to go when/if we add more code

G-Rath commented on src/dateRange.ts:
```diff
@@ -0,0 +1,19 @@
+import { jsDateToDateTime } from './dateMath';
+
+export const utcWeekRange = (on: Date = new Date()): [Date, Date] => {
```
I would [label the items in this tuple](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-0.html#labeled-tuple-elements):

```suggestion
export const utcWeekRange = (on: Date = new Date()): [startOfWeek: Date, endOfWeek: Date] => {
```

G-Rath commented on src/dateRange.ts:
```diff
@@ -0,0 +1,19 @@
+  ];
+};
+
+export const utcDayRange = (on: Date = new Date()): [Date, Date] => {
```
I would [label the items in this tuple](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-0.html#labeled-tuple-elements):

```suggestion
export const utcDayRange = (on: Date = new Date()): [startOfDay: Date, endOfDay: Date] => {
```

G-Rath commented on src/dateMath.ts:
```diff
@@ -0,0 +1,8 @@
+// We always set an explicit timezone on our Luxon `DateTime` objects so that
+// they never fall back to the system timezone. Falling back to the system
+// timezone can give different results depending on the system.
+export const jsDateToDateTime = (jsDate: Date): DateTime => {
```
🐘 I'd just call this `toDateTime` because the fact that it takes a "js date" is expressed via the types - that also makes it a bit easier if we later decided we wanted to support e.g. a string too

G-Rath commented on src/projectworks.ts:
```diff
@@ -0,0 +1,95 @@
+      return false;
+    }
+
+    if (!('LeaveID' in maybeLeaves[0])) {
```
TypeScript is bailing out because `unknown` + `Array.isArray` results in `any[]`, so it's not telling you that your previous `typeof !== 'object'` check will pass if the element is `null`, which'll then error here - so you want to add a guard for that (you can actually just change the condition on line 56 to be `maybeLeaves[0]` since that'll return `undefined` if there is no item)

eoinkelly replied:
Gotcha. I have also realised that `Array.isArray` covers the `null|undefined` cases too. I have refactored this whole thing.

G-Rath commented on src/projectworks.ts:
```diff
@@ -0,0 +1,95 @@
+};
+
+export const fetchLeaves = async (
+  params: Params
```
I'd just use `Record` instead of creating a dedicated interface since you're only using it here:

```suggestion
  params: Record<string, string>
```

G-Rath replied:
I'd just use `Record` instead of creating a dedicated interface since you're only using it here:

```suggestion
  params: Record<string, string>
```

G-Rath commented on test/src/dateRange.spec.ts:
```diff
@@ -0,0 +1,61 @@
+};
+
+describe('utcWeekRange', () => {
+  it('returns Date objects representing the start and end of the week containing the given Date', () => {
```
These should ideally be done using [`it.each`](https://jestjs.io/docs/api#testeachtablename-fn-timeout)

G-Rath commented on src/publicHolidayReminder.ts:
```diff
@@ -1,5 +1,12 @@
+  CalendarResponse,
+  DateWithTimeZone,
+  VEvent,
+  async
```
I think it would be best to alias this to avoid confusing it with the `async` keyword:

```suggestion
  async as ical
```

G-Rath commented on src/publicHolidayReminder.ts:
```diff
@@ -41,33 +48,53 @@ const toDateStamp = date => {
+  date: Date
+): MyEvent[] => {
+  return Object.values(calendarResponse)
+    .filter(rawEvent => {
+      return (
+        isVEvent(rawEvent) && toDateStamp(rawEvent.start) === toDateStamp(date)
+      );
+    })
     .map(rawEvent => {
+      const vEvent = rawEvent as VEvent; // safe to cast here because we ran isVEvent in filter() above
```
You can eliminate the cast by making the `filter` an explicit type guard:

```suggestion
    .filter((rawEvent): rawEvent is VEvent => {
      return (
        isVEvent(rawEvent) && toDateStamp(rawEvent.start) === toDateStamp(date)
      );
    })
    .map(rawEvent => {
      const vEvent = rawEvent;
```

G-Rath commented on src/publicHolidayReminder.ts:
```diff
@@ -41,33 +48,53 @@ const toDateStamp = date => {
+  date: Date
+): MyEvent[] => {
+  return Object.values(calendarResponse)
+    .filter(rawEvent => {
+      return (
+        isVEvent(rawEvent) && toDateStamp(rawEvent.start) === toDateStamp(date)
+      );
+    })
     .map(rawEvent => {
+      const vEvent = rawEvent as VEvent; // safe to cast here because we ran isVEvent in filter() above
```
You can eliminate the cast by making the `filter` an explicit type guard:

```suggestion
    .filter((rawEvent): rawEvent is VEvent => {
      return (
        isVEvent(rawEvent) && toDateStamp(rawEvent.start) === toDateStamp(date)
      );
    })
    .map(rawEvent => {
      const vEvent = rawEvent;
```

eoinkelly replied:
Ah ok. Somehow using a type guard function to do other things also passed me by. Thanks!

G-Rath replied:
`filter` is a classic gotcha, because by definition it always takes a typeguard but for safety reasons TypeScript is not allowed to infer typeguards because otherwise stuff could really get off the rails in subtle ways - a good rule of thumb is to always make the `filter` predicate function a type guard

G-Rath commented on src/handlers.ts:
```diff
@@ -0,0 +1,23 @@
+import * as publicHolidayReminder from './publicHolidayReminder';
+import type { NotifierResult } from './types';
+
+export const weeklyReport = async (): Promise<NotifierResult> => {
```
🐘 Ideally it would be good to use the types from `@types/aws-lambda` but that might be a bit of extra work

eoinkelly replied:
I think it's worth adding them for completeness

G-Rath commented on serverless.yml:
```diff
@@ -1,36 +1,53 @@
+configValidationMode: error
+plugins:
+  - serverless-plugin-scripts
+  - serverless-dotenv-plugin
```
We should be able to move off this plugin with `useDotenv` which is built-in to serverless now

eoinkelly replied:
I read 

> The framework only loads environments variables locally and does not pass them to your function's environment

in https://www.serverless.com/framework/docs/environment-variables and assumed that `useDotEnv` won't let us deprecate (I didn't test it however)

G-Rath replied:
That is correct which means it's not a drop-in replacement - so instead you want to define the env variables in the `serverless.yml` config like we do [here](https://github.com/ackama/serverless-aws-template/blob/main/serverless.yml#L36-L38); imo this is better because (among other things) you can check what env variables are used by a lambda without having to hunt around the codebase, you can provide env variables to specific functions only (instead of all of them), and we can enforce an env variable be present or otherwise give it a default value.

eoinkelly replied:
ah ok I misread it. I have updated it and it seems to work fine in staging

G-Rath reviewed (changes requested):
Looks like a good start - I've got a number of comments to discuss :)

eoinkelly committed:
Update src/dateRange.ts

Co-authored-by: Gareth Jones <gareth.jones@ackama.com>

eoinkelly committed:
Update src/dateRange.ts

Co-authored-by: Gareth Jones <gareth.jones@ackama.com>

eoinkelly committed:
Update src/publicHolidayReminder.ts

Co-authored-by: Gareth Jones <gareth.jones@ackama.com>

eoinkelly committed:
Refactor

eoinkelly committed:
Merge branch 'move-to-ackama-sls-template' of github.com:ackama/projectworks-leave-notifications into move-to-ackama-sls-template

eoinkelly committed:
Update src/publicHolidayReminder.ts

Co-authored-by: Gareth Jones <gareth.jones@ackama.com>

eoinkelly committed:
Refactor based on PR feedback

eoinkelly committed:
More refactoring

eoinkelly commented on src/projectworks.ts:
```diff
@@ -0,0 +1,83 @@
+const PW_USERNAME = process.env.PROJECTWORKS_USERNAME;
+const PW_PASSWORD = process.env.PROJECTWORKS_PASSWORD;
+
+const isObjectWithPropName = (ob: unknown, propName: string): boolean => {
+  if (!ob) {
+    return false;
+  }
+
+  if (typeof ob !== 'object') {
+    return false;
+  }
+
+  if (!(propName in ob)) {
+    return false;
+  }
+
+  return true;
+};
```
I'm tempted to replace this with `lodash` tbh. I know lodash in client-side code is tricky because it can be a big dependency but that wouldn't be an issue here. 

Built-in JS does quite a lot but there are gaps (like this) - how sad would you be if we pulled in lodash @G-Rath ?

G-Rath replied:
I was actually thankful you hadn't brought in `lodash` for this 😅

`lodash` is still a very big dependency which is not really maintained anymore, and you'd be pulling it in to cover two usages of a small function you've already written, which really you've only got because we've not properly typed out the PW API responses - so I'd rather we did that then bring in `lodash`

eoinkelly replied:
I have pulled these functions into a module and mimicked the lodash API because I think it's quite clean and composable for these kinds of tests and because I want a file I can easily paste into other apps (I seem to hit a "wishing I had lodash" point on every lambda I work on)

G-Rath commented on src/projectworks.ts:
```diff
@@ -0,0 +1,83 @@
+    return false;
+  }
+
+  if (maybeLeaves[0] && !isObjectWithPropName(maybeLeaves[0], 'LeaveID')) {
+    return false;
+  }
+
+  return true;
```
This is incorrect, since you're only saying it's not valid leaves if the first item exists _and_ it doesn't have a `LeaveID` property - you want this instead:

```suggestion
  if (isObjectWithPropName(maybeLeaves[0], 'LeaveID')) {
    return true;
  }

  return false;
```

eoinkelly replied:
Ah good catch. I'm not going to accept your suggestion because it would break the pattern of early `false` returns that I've used in other functions in this file but I'll sort the issue.

eoinkelly committed:
More refactoring

eoinkelly committed:
Refactor lowerdash

eoinkelly committed:
Add AWS lambda types to handler functions

eoinkelly committed:
Fix issue #24: erroring when sending 50+ slack blocks in a message

Split the blocks into chunks of 50.

eoinkelly committed:
Remove unneeded console.log

eoinkelly committed:
Remove serverless-dotenv-plugin

* Remove serverless-dotenv-plugin because we don't need it
* Remove forgotten date-fns-tz package
* Set server TZ to Pacific/Auckland - this is just a fallback in case we
  ever have code which depends on the local system TZ

G-Rath commented on src/projectworks.ts:
```diff
@@ -0,0 +1,64 @@
+export const fetchUser = async (userId: number): Promise<ProjectWorksUser> => {
+  const rawUser = await apiGet(`/api/v1.0/Users/${userId}`);
+
+  const user: unknown = await rawUser.json();
+
+  if (isValidUser(user)) {
+    return user;
+  }
+
+  throw new Error('Failed to get User');
```
I know we've just gone back and forward refactoring these validation checks, but I'm curious to why you're actually doing them in the first place vs just defining an interface with the appropriate properties? I'd be pretty concerned if the public PW API is so unstable...

eoinkelly replied:
My goal was "Take the JS that already works and just make it TS with as few changes as possible" so I've ended up here through a number of rounds of 1) make TS happy, 2) refactor, 3) goto 1.

That being said, I have time now and I'm keen to learn a bit. I should check the response for HTTP 400/500 (which I will do) which leaves me satisfying the types. I'm not sure I follow your question about defining an interface with appropriate properties. 

My understanding is that I need to get from a `Promise<any>` (as returned by `.json()` to a `ProjectWorksUser` (which I have already defined). 

From my reading of the TS docs I can either just use a [type assertion](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#type-assertions) to do this e.g.

```ts
const user = (await response.json()) as ProjectWorksUser; // json() returns Promise<any>
```

or use [type narrowing](https://www.typescriptlang.org/docs/handbook/2/narrowing.html). Narrowing seemed more robust so I followed that path. That being said, re-reading those docs today, I see that type narrowing is really more excluding types from a given union type rather than "this blob of JSON isn't what I expected".

Is this a place to just use the type assertion?

G-Rath replied:
Yup, what you've said pretty much hits the nail on the head - validating the response isn't strictly wrong, but one of the implications of a publicly documented API is a level of stability so that we can forgo the extra complexity of that validation safely;  so casting is the way to go here imo

G-Rath commented on test/src/dateRange.spec.ts:
```diff
@@ -0,0 +1,60 @@
+});
+
+describe('utcDayRange', () => {
+  it('returns Date objects representing the start and end of the day containing the given Date', () => {
```
This too should also be done with `it.each`

eoinkelly replied:
doh, thanks.

G-Rath commented on src/handlers.ts:
```diff
@@ -0,0 +1,28 @@
+import { Notifier } from './notifier';
+import * as publicHolidayReminder from './publicHolidayReminder';
+
+export const weeklyReport = async (
```
🐘 you can actually type these as `ScheduledHandler` to avoid having to define the unused params & return (i.e `export const weeklyReport: ScheduledHandler = async ()  => { ... }`)

eoinkelly replied:
That's neat, will do.

G-Rath reviewed (commented):
Looks good - main question I'd like to discuss is why you're actually doing validation on the PW API response rather than just typing it per their spec

eoinkelly committed:
Refactorings based on PR feedback

eoinkelly committed:
Clean up types in projectworks

eoinkelly committed:
Add .prettierignore to silence prettier warnings about LICENSE

G-Rath commented on src/projectworks.ts:
```diff
@@ -0,0 +1,45 @@
+  headers.append(
+    'Authorization',
+    `Basic ${Buffer.from(
+      `${PW_USERNAME as string}:${PW_PASSWORD as string}`
```
🐘 you can tell TypeScript that these variables will always be set in the environment by putting this at the top level:

```ts
declare global {
  namespace NodeJS {
    interface ProcessEnv {
      PROJECTWORKS_USERNAME: string;
      PROJECTWORKS_PASSWORD: string;
    }
  }
}
```

eoinkelly committed:
Built-in serverless useDotEnv will not load .env if .env.STAGENAME exists

This change also renames the example stage .envs to have `.example` at
the end which makes them easier to rename.

G-Rath commented on src/lowerDash.ts:
```diff
@@ -0,0 +1,25 @@
+// Utility functions inspired by (but not exactly matching) lodash API
+// https://lodash.com/docs/
+
+export const isArray = (thing: unknown): thing is unknown[] => {
```
Except for `chunk`, these are not being used anymore so we should axe them

eoinkelly committed:
Improve README

eoinkelly committed:
Remove unused functions

eoinkelly committed:
Fix error in deployment instructions
